core/src/main/java/org/sonarsource/sonarlint/core/tracking/Tracker.java

```diff
@@ -0,0 +1,416 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.tracking;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * Match and track a collection of issues.
+ *
+ * @param <R> type of the "raw" trackables that are in the incoming collection
+ * @param <B> type of the base trackables that are in the current collection
+ */
+public class Tracker<R extends Trackable, B extends Trackable> {
+
+  public Tracking<R, B> track(TrackableSupplier<R> rawTrackableSupplier, TrackableSupplier<B> baseTrackableSupplier) {
+    Tracking<R, B> tracking = new Tracking<>(rawTrackableSupplier, baseTrackableSupplier);
+
+    // 1. match issues with same rule, same line and same text range hash, but not necessarily with same message
+    match(tracking, LineAndTextRangeHashKeyFactory.INSTANCE);
+
+    // 2. match issues with same rule, same message and same text range hash
+    match(tracking, TextRangeHashAndMessageKeyFactory.INSTANCE);
+
+    // 3. match issues with same rule, same line and same message
+    match(tracking, LineAndMessageKeyFactory.INSTANCE);
+
+    // 4. match issues with same rule and same text range hash but different line and different message.
+    // See SONAR-2812
+    match(tracking, TextRangeHashKeyFactory.INSTANCE);
+
+    // 5. match issues with same rule, same line and same line hash
+    match(tracking, LineAndLineHashKeyFactory.INSTANCE);
+
+    // 6. match issues with same rule and same same line hash
+    match(tracking, LineHashKeyFactory.INSTANCE);
+
+    // 7. match issues with same server issue key
+    match(tracking, ServerIssueSearchKeyFactory.INSTANCE);
+
+    return tracking;
+  }
+
+  private void match(Tracking<R, B> tracking, SearchKeyFactory factory) {
+    if (tracking.isComplete()) {
+      return;
+    }
+
+    Map<SearchKey, List<B>> baseSearch = new HashMap<>();
+    for (B base : tracking.getUnmatchedBases()) {
+      SearchKey searchKey = factory.apply(base);
+      if (!baseSearch.containsKey(searchKey)) {
+        baseSearch.put(searchKey, new ArrayList<>());
+      }
+      baseSearch.get(searchKey).add(base);
+    }
+
+    for (R raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.apply(raw);
+      Collection<B> bases = baseSearch.get(rawKey);
+      if (bases != null && !bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)